libraries/ESP8266WiFi/src/BearSSLHelpers.cpp

```diff
@@ -0,0 +1,830 @@
+    xcs[1].data_len = 0;
+    *num = 1;
+    return xcs;
+  }
+
+  pos = decode_pem(buff, len, &num_pos);
+  if (!pos) {
+    return nullptr;
+  }
+  for (u = 0; u < num_pos; u ++) {
+    if (!strcmp_P(pos[u].name, PSTR("CERTIFICATE")) || !strcmp_P(pos[u].name, PSTR("X509 CERTIFICATE")))
+    {
+      br_x509_certificate xc;
+      xc.data = pos[u].data;
+      xc.data_len = pos[u].data_len;
+      pos[u].data = nullptr; // Don't free the data we moved to the xc vector!
+      cert_list.push_back(xc);
+    }
+  }
+  for (u = 0; u < num_pos; u ++) {
+    free_pem_object_contents(&pos[u]);
+  }
+  free(pos);
+
+  if (cert_list.size() == 0) {
+    return nullptr;
+  }
+  *num = cert_list.size();
+  dummy.data = nullptr;
+  dummy.data_len = 0;
+  cert_list.push_back(dummy);
+  xcs = (br_x509_certificate*)malloc(cert_list.size() * sizeof(*xcs));
+  if (!xcs) {
+    for (size_t i=0; i<cert_list.size(); i++) {
+      free(cert_list[i].data); // Clean up any captured data blobs
+    }
+    return nullptr;
+  }
+  memcpy(xcs, &cert_list[0], cert_list.size() * sizeof(br_x509_certificate));
+  // XCS now has [].data pointing to the previously allocated blobs, so don't
+  // want to free anything in cert_list[].
+  return xcs;
+}
+
+void free_certificates(br_x509_certificate *certs, size_t num)
+{
+  if (certs) {
+    for (size_t u = 0; u < num; u ++) {
+      free(certs[u].data);
+    }
+    free(certs);
+  }
+}
+
+static public_key *decode_public_key(const unsigned char *buff, size_t len)
+{
+  std::unique_ptr<br_pkey_decoder_context> dc(new br_pkey_decoder_context); // auto-delete on exit
+  if (!dc.get()) return nullptr;
+
+  public_key *pk = nullptr;
+
+  br_pkey_decoder_init(dc.get());
+  br_pkey_decoder_push(dc.get(), buff, len);
+  int err = br_pkey_decoder_last_error(dc.get());
+  if (err != 0) {
+    return nullptr;
+  }
+
+  const br_rsa_public_key *rk = nullptr;
+  switch (br_pkey_decoder_key_type(dc.get())) {
+    case BR_KEYTYPE_RSA:
+      rk = br_pkey_decoder_get_rsa(dc.get());
+      pk = (public_key*)malloc(sizeof *pk);
+      if (!pk) {
+        return nullptr;
+      }
+      pk->key_type = BR_KEYTYPE_RSA;
+      pk->key.rsa.n = (uint8_t*)malloc(rk->nlen);
+      pk->key.rsa.e = (uint8_t*)malloc(rk->elen);
+      if (!pk->key.rsa.n || !pk->key.rsa.e) {
+        free(pk->key.rsa.n);
+        free(pk->key.rsa.e);
+        free(pk);
+        return nullptr;
+      }
+      memcpy(pk->key.rsa.n, rk->n, rk->nlen);
+      pk->key.rsa.nlen = rk->nlen;
+      memcpy(pk->key.rsa.e, rk->e, rk->elen);
+      pk->key.rsa.elen = rk->elen;
+      return pk;
+
+    case BR_KEYTYPE_EC:
+      // TODO - not parsed yet in .T0 file
+      return nullptr;
+
+    default:
+      return nullptr;
+  }
+}
+
+void free_public_key(public_key *pk)
+{
+  if (pk) {
+    if (pk->key_type == BR_KEYTYPE_RSA) {
+      free(pk->key.rsa.n);
+      free(pk->key.rsa.e);
+    } else if (pk->key_type == BR_KEYTYPE_EC) {
+      // TODO - EC public keys not implemented
```

Comment:
Will this be done?

Reply:
I don't know how to generate or parse an EC key in ASN.1 format, so right now **I** won't be doing it, but someone can at some point in the future (or I'll figure it out). 

This is definitely not a common use case (EC keys), so not a big problem I'd say.